Make trait refs & assoc ty paths properly induce trait object lifetime defaults

fmease opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/129543)*
<!-- homu-ignore:end -->

## Trait Object Lifetime Defaults

### Primer & Definitions

You could read [this section](https://doc.rust-lang.org/reference/lifetime-elision.html#default-trait-object-lifetimes) in the Reference but it has several issues (see rust-lang/reference#1407). Here's a small explainer by me that only mentions the parts relevant to this PR:

Basically, given `dyn Trait` (≠ `dyn Trait + '_`) we want to deduce its *trait object lifetime bound* from context without relying on normal region inference as we might not be in a body[^1]. The "context" means the closest – what I call – *(eligible) container* `C<X0, …, Xn>` that wraps this trait object type. A *container* is to be understood as a use site of a "parametrized definition" (more general than type constructors). Currently *eligible* are ADTs, type aliases, traits and enum variants.

So if we have `C<dyn Trait>` (e.g., `&'r dyn Trait` or `Struct<'r, dyn Trait>`), `D<C<dyn Trait>>` or `C<N<dyn Trait>>` (e.g., `Struct<'r, (dyn Trait,)>`), we use the explicit[^2] outlives-bounds on the corresponding type parameter of `C` to determine the trait object lifetime bound. Here, `C` & `D` denote (eligible) containers and `N` denotes a generic type that is **not** an eligible container. E.g., given `struct Struct<'a, T: 'a + ?Sized>(…);`, we elaborate `Struct<'r, dyn Trait>` to `Struct<'r, dyn Trait + 'r>`.

Finally, we call lifetime bounds used as the default for *constituent* trait object types of an eligible container `C` the *trait object lifetime defaults* (*induced by* `C`). These defaults may of course end up getting shadowed in parts of the type by the defaults induced by any inner eligible containers.

### Changes Made

**These changes are theoretically breaking**.

1. Make *resolved* associated type paths / projections eligible containers.
   * `<Y0 as TraitRef<X0, …, Xn>>::AssocTy<Y1, …, Ym>` now induces *trait object lifetime defaults* for constituents `Y0` to `Ym` (`TraitRef` is considered a separate container, see also list item **(3)**).
   * Notably, for the self type `Y0` of (resolved) projections we now look at the bounds on the `Self` type param of the relevant trait (e.g., given `trait Outer<'a>: 'a { type Proj; }` or `trait Outer<'a> where Self: 'a { type Proj; }` we elaborate `<dyn Inner as Outer<'r>>::Proj` to `<dyn Inner + 'r as Outer<'r>>::Proj`).
   * Example breakages:
     ```rs
     trait Outer<'a> { type Ty<T: ?Sized + 'a>; }
     impl<'a> Outer<'a> for () { type Ty<T: ?Sized + 'a> = &'a T; }
     trait Inner {}
 
     fn f<'r>(x: <() as Outer<'r>>::Ty<dyn Inner>) {
     //                                ~~~~~~~~~
     //                                this branch:  dyn Inner + 'r       (due to bound `'a` on `T`)
     //                                stable/main:  dyn Inner + 'static  (due to item signature fallback)
         let _: <() as Outer<'r>>::Ty<dyn Inner + 'static> = x;
     //         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     //         this branch:  error: lifetime may not live long enough // `'r` must outlive `'static`
     //         stable/main:  OK
     }
     ```
     ```rs
     trait Outer { type Ty; }
     trait Inner {}

     impl<'a> Outer for dyn Inner + 'a { type Ty = &'a (); }

     fn f<'r>(x: *mut &'r <dyn Inner as Outer>::Ty) {
     //                    ~~~~~~~~~
     //                    this branch:  dyn Inner + 'static  (due to lack of bounds on `Outer`; the assoc type path shadows the default induced by the type ctor `&`)
     //                    stable/main:  dyn Inner + 'r       (due to bound `'a` on `T` in (pseudo) `builtin type &<'a, T: 'a + ?Sized>;`)
         let _: *mut &'r <dyn Inner + 'r as Outer>::Ty = x;
     //         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     //         this branch:  error: lifetime may not live long enough // `'r` must outlive `'static`
     //         stable/main:  OK
     }
     ```
2. In *type-relative* paths `Y0::Name<Y1, …, Ym>` consider the trait object lifetime default **indeterminate**
   * Meaning if we're in an "item context" / "item signature" / "non-body" (& the principal trait isn't bounded by any outlives-bounds which would take precedence over the default) we will **reject** any implicit trait object lifetime bounds that would take on that default
   * Reason: Limitations of the current implementation which can't be easily overcome
     * RBV (which resolves trait object lifetime defaults by recursing into the local crate "in one sitting") would require the resolution of *type-relative* paths in order to look up the generics but these paths are only resolved in HIR ty lowering (that can selectively lower local items) which depends on the results of RBV (cyclic dependency!)
     * While one might be able to resolve type-relative paths in RBV in an ad-hoc fashion, it would require a lot of duplication with HIR ty lowering and its impl would be very brittle (RTN does something like that in RBV but we require a more sophisticated resolver)
     * I did attempt that but it got too gnarly and brittle and would've likely been incomplete anyway
     * See also [this GH thread](https://github.com/rust-lang/rust/pull/129543#discussion_r2079677600)
     * See also [#t-types/meetings > 2025-09-16 weekly @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/326132-t-types.2Fmeetings/topic/2025-09-16.20weekly/near/539889059)
   * This should still be maximally forward compatible and allow us to implement the desired behavior in the future.
   * Example breakage:
     ```rs
     trait Outer { type Ty<'a, T: 'a + ?Sized>; }
     trait Inner {}
 
     fn f<'r, T: Outer>(x: T::Ty<'r, dyn Inner>) {}
     //                              ~~~~~~~~~
     //                              this branch:  error: indeterminate  (reservation)
     //                              stable/main:  dyn Inner + 'static   (due to item signature fallback)
     ```
3. Fixes trait object lifetime defaults inside trait refs `TraitRef<X0, …, Xn>` (this fell out from the previous changes). They used to be completely broken due to a nasty off-by-one error for not accounting for the implicit `Self` type param of traits which lead to cases like
   * `Outer<'r, dyn Inner>` (with `trait Outer<'a, T: 'a + ?Sized> {}`) getting rejected as *indeterminate* (it tries to access a *lifetime* at index 1 instead 0) ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=0069c89b2313f0f447ff8b6f7de9adfa)) 
   * `Outer<'r, 's, dyn Inner>` (with `trait Outer<'a, 'b, T: 'a + ?Sized> {}`) elaborating `dyn Inner` to `dyn Inner + 's` instead of `dyn Inner + 'r`(!) which subsequently gets rejected of course since `'s` isn't known to outlive `'r` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=9c521165e0ac0d868a8087cd7ca861fe))
   * The same applies to trait *alias* refs (feature `trait_alias`)
   * Example breakage:
     ```rs
     trait Outer<'a, 'b, T: 'a + ?Sized> {}
     trait Inner {}
 
     struct F<'r, T>
     where
         T: Outer<'r, 'static, dyn Inner>
     //                        ~~~~~~~~~
     //                        this branch:  dyn Inner + 'r       (correctly mapping `'a` => `'r`)
     //                        stable/main:  dyn Inner + 'static  (incorrectly mapping `'a` => `'static` due to off-by-one)
     {
         g: G<'r, T>,
     //     ~~~~~~~~
     //     this branch:  error: mismatched types
     //                          expected: `Outer<'r, 'static, (dyn Inner + 'static)>`
     //                             found: `Outer<'r, 'static, (dyn Inner + 'r)>`
     //     stable/main:  OK
     }
 
     struct G<'r, T>(&'r (), T)
     where
         T: Outer<'r, 'static, dyn Inner + 'static>;
     ```
4. In associated type binding `TraitRef<AssocTy<X0, …, Xn> = Y>` consider the trait object lifetime default **indeterminate** (in `X0`, …, `Xn` and `Y`) if `X0`, …, `Xn` contains any lifetime arguments.
   * Meaning if we're in an item context (& the principal trait isn't bounded) we will **reject** any implicit trait object lifetime bounds that would take on that default
   * This reserves us the right to (1) take into account the *item bounds* of `AssocTy` in the future when computing the default for `Y` (2) take into account the parameter bounds of `AssocTy` in the future when computing the defaults for `X0`, …, `Xn`.
   * This extends a preexisting hack that – given `TraitRef<X0, …, Xn, AssocTy<Y0, …, Ym> = Z>` – treats the default indeterminate in `Y0`, …, `Ym` and `Z` if `X0`, …, `Xn` contains any lifetime arguments. 
   * Rephrased, this hack / reservation previously didn't account for GAT args, only trait ref args, which is insufficient
   * See also [this GH comment of mine](https://github.com/rust-lang/rust/issues/115379#issuecomment-3751616772)
   * Example breakages:
     ```rs
     trait Outer { type Ty<'a>: ?Sized; }
     trait Inner {}

     fn f<'r>(_: impl Outer<Ty<'r> = dyn Inner>) {}
     //                              ~~~~~~~~~
     //                              this branch:  error: indeterminate  (reservation)
     //                              stable/main:  dyn Inner + 'static   (forced)
     ```
     ```rs
     trait Outer { type Ty<'a, T: ?Sized + 'a>; }
     trait Inner {}

     fn f<'r>(_: impl Outer<Ty<'r, dyn Inner> = ()>) {}
     //                            ~~~~~~~~~
     //                            this branch:  error: indeterminate  (reservation)
     //                            stable/main:  dyn Inner + 'static   (forced)
     ```

#### Motivation

Both trait object lifetime default RFCs ([599](https://rust-lang.github.io/rfcs/0599-default-object-bound.html) and [1156](https://rust-lang.github.io/rfcs/1156-adjust-default-object-bounds.html)) never explicitly specify what constitutes a — what I call — *(eligible) container* but it only makes sense to include anything that can be parametrized by generics and can be mentioned in places where we don't perform full region inference … like associated types. So it's only *consistent* to make this change.

#### Breakages

**These changes are theoretically breaking** because they can lead to different trait object lifetime bounds getting deduced compared to main which is obviously user observable. Moreover, we're now explicitly rejecting implicit trait object lifetime bounds inside type-relative paths (excl. the self type) and on the RHS of assoc type bindings if the assoc type has lifetime params.
However, the latest crater run found 0 non-spurious regressions (see [here](https://github.com/rust-lang/rust/pull/129543#issuecomment-3787331877) and [here](https://github.com/rust-lang/rust/pull/129543#issuecomment-3795138538)).

---

Fixes rust-lang/rust#115379.
Fixes rust-lang/rust#140710.
Fixes rust-lang/rust#141997.


[^1]: If we *are* in a body we do use to normal region inference as a fallback.
[^2]: Indeed, we don't consider implied bounds (inferred outlives-bounds).

fmease commented:
@bors try

bors commented:
:hourglass: Trying commit 8bfcd86acbecdac23e6236f6ab0e1732448d4f8f with merge 24cd45d7714ba43afd4b8e62fb677b069b21c4a5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8bfcd86acbecdac23e6236f6ab0e1732448d4f8f","merge_sha":"24cd45d7714ba43afd4b8e62fb677b069b21c4a5"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10543682495/job/29213223696)
Build commit: 24cd45d7714ba43afd4b8e62fb677b069b21c4a5 (`24cd45d7714ba43afd4b8e62fb677b069b21c4a5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10543682495/job/29213223696"},"merge_sha":"24cd45d7714ba43afd4b8e62fb677b069b21c4a5"} -->

fmease commented:
@craterbot check

fmease committed:
Tweak `#[rustc_dump_object_lifetime_defaults]`

* print `Empty` as it's called in code, otherwise it's unnecessarily confusing
* go through the middle::ty Generics instead of the HIR ones, so we can dump the
  default for the implicit `Self` type parameter of traits, too

fmease committed:
Don't needlessly search for already-found HIR generic param

fmease committed:
Make resolved assoc ty paths induce a trait object lifetime default for their self ty

fmease commented:
@bors try

bors commented:
:hourglass: Trying commit bb92e569fcadb82c32674bd5a5a7bc0421d48402 with merge 0e6f83462d087486e5a969d6c90de0c91cf2d9ee...
<!-- homu: {"type":"TryBuildStarted","head_sha":"bb92e569fcadb82c32674bd5a5a7bc0421d48402","merge_sha":"0e6f83462d087486e5a969d6c90de0c91cf2d9ee"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/14563962357/job/40850663216) [(plain)](https://github.com/rust-lang-ci/rust/commit/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/checks/40850663216/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.url=https://github.com/rust-lang/rust
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
   Compiling rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
   Compiling rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
   Compiling rustc_privacy v0.0.0 (/checkout/compiler/rustc_privacy)
   Compiling rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
error[E0609]: no field `res` on type `&&rustc_hir::Lifetime`
    --> compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:1107:37
     |
1107 |                 set.insert(lifetime.res);
     |                                     ^^^ unknown field
     |
     = note: available fields are: `hir_id`, `ident`, `kind`, `is_anon_in_path`

[RUSTC-TIMING] rustc_sanitizers test:false 4.718
For more information about this error, try `rustc --explain E0609`.
[RUSTC-TIMING] rustc_hir_analysis test:false 5.584
error: could not compile `rustc_hir_analysis` (lib) due to 1 previous error
---
Caused by:
    Command RUST_BACKTRACE=full python3 /checkout/x.py build --target x86_64-unknown-linux-gnu --host x86_64-unknown-linux-gnu --stage 2 library/std --rust-profile-generate /tmp/tmp-multistage/opt-artifacts/rustc-pgo --set llvm.thin-lto=false --set llvm.link-shared=true [at /checkout/obj] has failed with exit code Some(1)

Stack backtrace:
   0: <anyhow::Error>::msg::<alloc::string::String>
             at /rust/deps/anyhow-1.0.97/src/backtrace.rs:27:14
   1: <opt_dist::exec::CmdBuilder>::run
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/exec.rs:80:17
   2: <opt_dist::exec::Bootstrap>::run
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/exec.rs:181:9
   3: opt_dist::execute_pipeline::{closure#1}::{closure#0}
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/main.rs:230:13
   4: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}::{closure#0}, ()>
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/timer.rs:111:22
   5: opt_dist::execute_pipeline::{closure#1}
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/main.rs:219:9
   6: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}, opt_dist::training::RustcPGOProfile>
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/timer.rs:111:22
   7: opt_dist::execute_pipeline
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/main.rs:216:29
   8: opt_dist::main
             at /rustc/0e6f83462d087486e5a969d6c90de0c91cf2d9ee/src/tools/opt-dist/src/main.rs:416:18
   9: <fn() -> core::result::Result<(), anyhow::Error> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/core/src/ops/function.rs:250:5
  10: std::sys::backtrace::__rust_begin_short_backtrace::<fn() -> core::result::Result<(), anyhow::Error>, core::result::Result<(), anyhow::Error>>
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/sys/backtrace.rs:152:18
  11: std::rt::lang_start::<core::result::Result<(), anyhow::Error>>::{closure#0}
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/rt.rs:199:18
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/core/src/ops/function.rs:284:13
  13: std::panicking::try::do_call
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/panicking.rs:589:40
  14: std::panicking::try
             at /rustc/45165c82a4c5315ff52c391ad138f41ff40b52d8/library/std/src/panicking.rs:552:19
```

</details>

fmease commented:
@bors try

bors commented:
:hourglass: Trying commit 514c3642e776ddcfa88a476d4fc69c1e5a093536 with merge 9b14abcd2fc6076c7fc2047b85ebb91e2302ff6b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"514c3642e776ddcfa88a476d4fc69c1e5a093536","merge_sha":"9b14abcd2fc6076c7fc2047b85ebb91e2302ff6b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14582801162/job/40906963536)
Build commit: 9b14abcd2fc6076c7fc2047b85ebb91e2302ff6b (`9b14abcd2fc6076c7fc2047b85ebb91e2302ff6b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14582801162/job/40906963536"},"merge_sha":"9b14abcd2fc6076c7fc2047b85ebb91e2302ff6b"} -->

fmease commented:
@craterbot check

compiler-errors commented:
@craterbot retry-report

fmease commented:
I've opened a Zulip topic, so we don't need to spam this issue :)

([#t-infra > Crater failed to build report twice @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Crater.20failed.20to.20build.20report.20twice/near/513842815)

Mark-Simulacrum commented:
@craterbot retry-report

The last two failures were spurious timeouts connecting to the sqlite db... which seems a bit odd (that's just a file open, I think?). Let's try one more time and if this also fails I'll have to dig into what's happening in more detail, perhaps pausing ongoing runs so there's less churn on the db while we load the report.

fmease commented:
Ready for review and also types FCP I'd say (cc https://github.com/rust-lang/rust/issues/115379#issuecomment-1733876704).

r? types

fmease commented on compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:
```diff
@@ -1035,54 +1119,54 @@ impl<'a, 'tcx> Visitor<'tcx> for BoundVarContext<'a, 'tcx> {
+    let (generics, bounds) = match tcx.hir_node_by_def_id(param_def_id) {
+        hir::Node::GenericParam(param) => match param.source {
+            hir::GenericParamSource::Generics => {
+                // FIXME: message
```
rm

fmease commented on compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:
```diff
@@ -887,11 +888,94 @@ impl<'a, 'tcx> Visitor<'tcx> for BoundVarContext<'a, 'tcx> {
+                        ),
+                        0,
+                    ) => Some((def_id, path.segments)),
+                    // Note: We don't need to care about definitions kinds that can have generics if they
```
```suggestion
                    // Note: We don't need to care about definition kinds that can have generics if they
```

fmease commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit ec131fe59f51dda526d23f3198fc5c6660b8ac42 with merge 99962d8169c938d94d1a9642217089b43363e569...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ec131fe59f51dda526d23f3198fc5c6660b8ac42","merge_sha":"99962d8169c938d94d1a9642217089b43363e569"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14892672437/job/41834454089)
Build commit: 99962d8169c938d94d1a9642217089b43363e569 (`99962d8169c938d94d1a9642217089b43363e569`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14892672437/job/41834454089"},"merge_sha":"99962d8169c938d94d1a9642217089b43363e569"} -->

rust-timer commented:
Queued 99962d8169c938d94d1a9642217089b43363e569 with parent 3ef8e64ce9f72ee8d600d55bc43b36eed069b252, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3ef8e64ce9f72ee8d600d55bc43b36eed069b252&end=99962d8169c938d94d1a9642217089b43363e569).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.6 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (99962d8169c938d94d1a9642217089b43363e569): [comparison URL](https://perf.rust-lang.org/compare.html?start=3ef8e64ce9f72ee8d600d55bc43b36eed069b252&end=99962d8169c938d94d1a9642217089b43363e569&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=3ef8e64ce9f72ee8d600d55bc43b36eed069b252&end=99962d8169c938d94d1a9642217089b43363e569&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [1.1%, 1.1%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [1.1%, 1.1%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=3ef8e64ce9f72ee8d600d55bc43b36eed069b252&end=99962d8169c938d94d1a9642217089b43363e569&stat=max-rss)
<details>
<summary>Results (primary 0.7%, secondary -0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [0.4%, 5.9%]   | 11    |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.9%, 1.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.2%, -0.5%] | 5     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-3.6%, -3.6%] | 1     |
| All ❌✅ (primary)                 | 0.7%  | [-1.2%, 5.9%]  | 16    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=3ef8e64ce9f72ee8d600d55bc43b36eed069b252&end=99962d8169c938d94d1a9642217089b43363e569&stat=cycles:u)
<details>
<summary>Results (primary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.5%, 0.8%]   | 5     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.3%, -0.5%] | 5     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-1.3%, 0.8%]  | 10    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=3ef8e64ce9f72ee8d600d55bc43b36eed069b252&end=99962d8169c938d94d1a9642217089b43363e569&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0% | [0.0%, 0.0%] | 8     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.0% | [0.0%, 0.0%] | 8     |
</details>

**Bootstrap**: 770.356s -> 769.507s (-0.11%)
**Artifact size**: 365.20 MiB -> 365.10 MiB (-0.03%)

oli-obk commented on tests/ui/object-lifetime/object-lifetime-default-gat-resolved.rs:
```diff
@@ -0,0 +1,33 @@
+    fn g<'r>(_: <() as Outer>::Ty<'r, dyn Inner>) {}
+}
+
+mod parent { // the object lifetime default comes from the parent generics
```
Also add a test with both self and parent lifetimes

fmease replied:
Do you mean one where the assoc ty is `type AssocTy<'own, T: ?Sized + 'own + 'parent>;` or `type AssocTy<'own, T: ?Sized + 'own, U: ?Sized + 'parent>;` or something else entirely?

oli-obk replied:
I had in mind just `type AssocTy<'own, T: ?Sized + 'own>` (basically identical to the first example, but showing that the extra lifetime doesn't throw it off), and ideally some sort of failure tests of cases that are using the wrong lifetime.

BoxyUwU commented:
Could you add an example to the "theoretically breaking" section with a little bit of prose explaining why it breaks? My understanding is that something like this would break:
```rust
trait Identity<'a>: 'a {
    type Assoc: ?Sized;
}

impl<'a, T: ?Sized + 'a> Identity<'a> for T {
    type Assoc = T;
}

trait Trait {}

fn foo<'b>(a: Box<<dyn Trait as Identity<'b>>::Assoc>) {
    let a: Box<dyn Trait + 'static> = a;
}
```
Right now it compiles because we don't use the `T: 'a` bound from the trait def to compute the dyn type's lifetime bound, but with this PR we will do so, winding up with `dyn Trait + 'a` instead of `dyn Trait + 'static`. Is that correct?

BoxyUwU commented on compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:
```diff
@@ -2158,6 +2197,83 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
+        speculative: bool,
+    ) -> Option<(Vec<ty::BoundVariableKind>, &'tcx ty::AssocItem)> {
+        // This mimics HIR ty lowering's `lower_assoc_path_shared`.
+        // FIXME: Duplicating efforts is not robust or sustainable/maintainable.
+        // Ideally, we'd simply obtain the resulting type-dependent defs from
+        // HIR ty lowering (not only in FnCtxts but also in ItemCtxts!).
```
I guess you said it yourself here :) 

Handling `TypeRelative` paths by duplicating hir ty lowering logic here seems unfortunate. Do you think it'd be reasonable to just continue with the stable behaviour for `TypeRelative` paths, this PR would still be a positive change for fully qualified aliases and trait refs.

Though, admittedly, I'm not sure how we'd ever make this work "properly". Propagating type dependent def resolutions from inside ty lowering into the internals here seems unlikely to happen.

Maybe we could have a special `resolve_dyn_type_lifetime_bound` query that returns whatever we resolved *and* how many type dependent defs we stepped through (i.e. some kind of `[unknown, unknown, 'a, unknown]`). Hir ty lowering could then track its own set of lifetime defaults coming from type dependent paths and pair that together to figure out what it should *actually* use.... Seems somewhat complicated :sweat_smile:

fmease replied:
Tbh, I'd find it really unfortunate if there were such an obscure behavior difference between resolved & type-relative paths! Esp. since my personal motivation for this PR is to 'streamline' the rules if object lifetime defaults by making them apply everywhere where one would expect them to.

For the longest time this branch simply rejected implicit object lifetime bounds in type-relative associated types — that I'd be fine with and crater also didn't complain about that. However, that'd regress `tests/ui/deriving/issue-89188-gat-hrtb.rs` / issue #89188.

fmease replied:
Note that `limited_resolve_type_relative_path` is already used by RTN, I only extracted the logic into a separate function, so I can use it. So even without this PR, in PR https://github.com/rust-lang/rust/pull/126651 (supporting `T::AssocTy0::AssocTy1` I need to update it.

BoxyUwU replied:
Ah okay :thinking: If it's pre-existing then I think it's fine :+1: even if long term would be nice to not have to do.

BoxyUwU reviewed (commented):
Feel generally positive towards the conceptual change to the type system here. It feels like a very straightforward "bug fix" change.

fmease commented on compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:
```diff
@@ -1859,6 +1851,110 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
+                }
+
+                let (depth, index) = param_to_depth_and_index(generics, self.tcx, param_def_id);
+                segments[segments.len() - depth - 1]
```
Right, I didn't think about parent-supplied lifetimes in type-relative paths. Thanks for catching this! Reduced:

```rs
trait Outer<'a> { type Ty<T: ?Sized + 'a>; }
impl<'a, X> Outer<'a> for X { type Ty<T: ?Sized + 'a> = (); }
trait Inner {}

fn f<'r, T: Outer<'r>>(a: T::Ty<dyn Inner>) {}

fn main() {}
```

BoxyUwU replied:
Much less messy than my example :laughing:

fmease commented:
> Could you add an example to the "theoretically breaking" section with a little bit of prose explaining why it breaks?

Will do!

> Right now it compiles because we don't use the `T: 'a` bound from the trait def to compute the dyn type's lifetime bound, but with this PR we will do so, winding up with `dyn Trait + 'a` instead of `dyn Trait + 'static`. Is that correct?

Basically yes. Before this PR, any bounds on type parameters of associated type paths were completely ignored when it comes to object lifetime defaults. And since ambient object lifetime defaults are scopes, we'd just use the 'next closest' scope:

* Looking at your example — `Box<<dyn Trait as Identity<'b>>::Assoc>` — **on master**, the projection does **not** induce any ambient object lifetime default and thus we'll use the one induced by `Box<…>` which is `'static` in this case. Under this PR the relevant scope structure looks like `<- Some('static) <- Some('b)` where `Some('b)` is induced by the projection which we gladly take as it's the innermost one.

lcnr reviewed (commented):
please add a test for
```rust
trait Trait<T> {
    type Assoc<'a>
    where
        T: 'a;
}
```
I would assume that we don't have a default for `(): Trait<dyn Inner>`. What about `(): Trait<dyn Inner, Assoc<'r> = ()>` or `<() as Trait<dyn Inner>>::Assoc<'r>`

bors commented:
:umbrella: The latest upstream changes (presumably #142255) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented:
> please add a test for
> 
> ```rust
> trait Trait<T> {
>     type Assoc<'a>
>     where
>         T: 'a;
> }
> ```
> 
> I would assume that we don't have a default for `(): Trait<dyn Inner>`. What about `(): Trait<dyn Inner, Assoc<'r> = ()>` or `<() as Trait<dyn Inner>>::Assoc<'r>`

In all three cases, the following holds:

1. `trait Inner {}` presumably isn't bounded by a lifetime
2. We're in an item ctxt / signature / non-body (in a where-clause to be precise)
   * Well, the third one, the projection, could be inside of a body, we don't know, but then it'd resolve to a fresh infer var, so that case isn't interesting
4. `Trait`'s `T` doesn't impose any lifetime requirements (we don't traverse into the associated items) aka it's `Empty`

We therefore default to `'static` which is the fallback in non-bodies for `Empty`.

I don't know, do you expect that in the third case we default to `'r`? I think I could make that work, just let me know.

fmease commented:
I've pushed a tmp commit to address https://github.com/rust-lang/rust/issues/115379#issuecomment-3751616772 which I'll need to crater. It's good to rerun crater anyway cuz the last one was almost a year ago.

@bors try

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/21010675894/job/60404575492) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/60404575492) [(plain)](https://github.com/rust-lang/rust/commit/c72baacce9d49d1c02842d80db185dd3977da9ca/checks/60404575492/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 11.850
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:1826:
         // in the trait ref `YY<...>` in `Item: YY<...>`.
         for constraint in generic_args.constraints {
             let scope = Scope::ObjectLifetimeDefault {
-                lifetime: if has_lifetime_parameters || constraint.gen_args.args.iter().any(|arg| matches!(arg, GenericArg::Lifetime(_))) {
+                lifetime: if has_lifetime_parameters
+                    || constraint
+                        .gen_args
+                        .args
+                        .iter()
+                        .any(|arg| matches!(arg, GenericArg::Lifetime(_)))
+                {
                     None
                 } else {
                     Some(ResolvedArg::StaticLifetime)
fmt: checked 6650 files
```

</details>

fmease commented:
@craterbot check

BoxyUwU commented:
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-129543-2/retry-regressed-list.txt p=1

fmease committed:
Rename `hir::GenericArgs::{num,has}_generic_params` to `*_generic_args`

fmease committed:
Render trait object lifetime defaults indeterminate in lifetime-instantiated assoc ty bindings

Namely, on the RHS and in the args of the bindings.

fmease committed:
Split out fn `compute_object_lifetime_defaults`

fmease commented on tests/ui/object-lifetime/object-lifetime-default-inferred.rs:
This test didn't actually test what it meant to test. I've therefore replaced it with a new one (`tests/ui/object-lifetime/object-lifetime-default-inferred-args.rs`). It was added in the initial `generic_arg_infer` PR.

It's meant to test that the presence of `_` in a generic arg list doesn't throw off the trait object lifetime default calculations which is a very valuable thing to test (hence me adding a replacement). Sadly, it doesn't test that.

First of all, the only places in this file that contain implicit trait object lifetime bounds that are in the mere *vicinity* of an inferred arg `_` are the two `&dyn Test`s at the bottom. However, these just get elab'ed to `&'?0 (dyn Test + '?1)` and `&'2 (dyn Test + '?3)` since the implicit inferred lifetimes passed to `&` aren't *named*, so the default is *indeterminate* meaning fresh region vars for both since we're in a body.

The two `_` passed as part of the fn calls have no influence whatsoever on the defaults chosen for the `dyn Test` "in random `as`-casts"; that's not how trait object lifetime defaulting works.

lcnr commented on tests/ui/object-lifetime/object-lifetime-default-trait-ref-lifetime-bound-on-assoc-ty.rs:
```diff
@@ -0,0 +1,27 @@
+//@ check-pass
+
+trait Trait<T: ?Sized> { type Assoc<'a> where T: 'a; }
+impl<T: ?Sized> Trait<T> for () { type Assoc<'a> = &'a T where T: 'a; }
```
is this impl needed?

lcnr commented on tests/ui/object-lifetime/object-lifetime-default-trait-ref.rs:
```diff
@@ -0,0 +1,34 @@
+// We deduce `dyn Inner + 'r` from bound `'a` on ty param `T` of trait `Outer`.
+fn assoc_ty_proj<'r>(_: <() as Outer<'r, dyn Inner>>::Ty) {}
+
+impl<'a, T: 'a + AbideBy<'a> + ?Sized> Outer<'a, T> for () { type Ty = &'a T; }
```
move this to trait definition?

lcnr commented on tests/ui/object-lifetime/object-lifetime-default-assoc-ty-binding-item-bounds.rs:
```diff
@@ -0,0 +1,37 @@
+
+fn is_static<T>(_: T) where T: 'static {}
+
+// FIXME: Ideally, we'd elaborate `dyn Bar` to `dyn Bar + 'static` instead of rejecting it.
```
do we? 😅 I am not even sure whether this is desirable or not.

lcnr replied:
3. ...induce a lifetime default that's determined by the item bounds
    * here that would be ~~`'static`~~*indeterminate* since OLD(`Foo::Item`) would be `Empty` which turns into ~~`'static`~~*indeterminate for associated type bindings* in item signatures

is what I'd expect

fmease replied:
Ah, I see. Well, as a matter of fact in all other contexts *Empty* just like *Static* leads to a `'static` lifetime default in item signatures, so that would be a bit inconsistent. I mean maybe it's fine if type args and assoc type bindings get treated differently, idk.

For comparison, `&'r Rc<dyn Trait>` gets elab'ed to `&'r Rc<dyn Trait + 'static>` in item signatures since `Rc`'s `T` is *Empty*. So it's neither indeterminate nor pass-through.

lcnr reviewed (commented):
When do we actually use trait object defaults in bodies? isn't it better to just always infer them unless we're in an `ItemCtxt`?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/129543/d00ba922591daa3d0a8b0f3cdf54c610097c24ea..2fec74f55d4ee73bc456f2d723166032acf8008c)*

fmease committed:
Introduce fn `eligible_container` & rename depth to rev seg index

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/24284316590/job/70911076186) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/70911076186) [(plain)](https://github.com/rust-lang/rust/commit/7600582499c690a511a55e346431e0d7e5ffba85/checks/70911076186/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error[E0023]: this pattern has 7 fields, but the corresponding tuple variant has 8 fields
    --> compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:1108:40
     |
1108 |             kind: hir::ItemKind::Trait(_, _, _, _, generics, bounds, _),
     |                                        ^  ^  ^  ^  ^^^^^^^^  ^^^^^^  ^ expected 8 fields, found 7
     |
    ::: compiler/rustc_hir/src/hir.rs:4542:5
     |
4542 |     Trait(
     |     ----- tuple variant defined here
4543 |         Constness,
     |         ---------
4544 |         IsAuto,
     |         ------
4545 |         Safety,
     |         ------
4546 |         &'hir ImplRestriction<'hir>,
     |         ---------------------------
4547 |         Ident,
     |         -----
4548 |         &'hir Generics<'hir>,
     |         --------------------
4549 |         GenericBounds<'hir>,
     |         -------------------
4550 |         &'hir [TraitItemId],
     |         ------------------- tuple variant has 8 fields
     |
help: use `_` to explicitly ignore each field
     |
1108 |             kind: hir::ItemKind::Trait(_, _, _, _, generics, bounds, _, _),
     |                                                                       +++

For more information about this error, try `rustc --explain E0023`.
[RUSTC-TIMING] rustc_hir_analysis test:false 5.620
error: could not compile `rustc_hir_analysis` (lib) due to 1 previous error
```

</details>

fmease committed:
Make resolved GAT paths induce trait object lifetime defaults for their own args (excl. self ty)

This automatically fixes trait object lifetime defaulting for trait refs
and trait alias refs, too.
These used to be broken because the index calculation for going from
middle generic args back to HIR ones didn't take into account the
implicit self ty param present on traits.

Moreover, in item signatures reject hidden trait object lifetime bounds
inside type-relative paths (excl. the self ty) on grounds of the default
being indeterminate.

lcnr commented on compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:
```diff
@@ -787,11 +787,10 @@ impl<'a, 'tcx> Visitor<'tcx> for BoundVarContext<'a, 'tcx> {
-                        self.resolve_object_lifetime_default(&*lifetime)
+                        // If the user doesn't write *anything*, we apply the
+                        // trait object lifetime defaulting rules.
+                        // E.g., `Box<dyn Debug>` becomes `Box<dyn Debug + 'static>`.
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/129543/bf4fbfb7a18d74e7cd8eef93af7329c58fbb5344..9b4931cdcb1b0d28854ea9d0cfd31df609a573bc)*

```suggestion
                        // E.g. `Box<dyn Debug>` becomes `Box<dyn Debug + 'static>`.
```

fmease replied:
A comma after "e.g." is a stylistic choice, very prevalent in American English (Chicago Style, AP Style, etc.). See also https://linguaholic.com/linguablog/comma-after-e-g.

lcnr commented:
gonna start a new FCP with this. see the PR description.

@rfcbot fcp merge types

fmease commented:
Maybe we need to cancel the old FCP first?

BoxyUwU commented:
@rfcbot cancel

BoxyUwU commented:
@rfcbot fcp merge types

BoxyUwU commented:
@rfcbot cancel

lcnr will have to do it it wont let me uncheck my box for some reason...

lcnr commented:
@rfcbot fcp cancel

lcnr commented:
@rfcbot fcp merge types

BoxyUwU commented:
unassigning myself because it seems like lcnr already did review. really nice PR description thanks fmease :)

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25669894983/job/75352180261) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/75352180261) [(plain)](https://github.com/rust-lang/rust/commit/b7cb40380ae014a8b0faaed3ade12a02786a8ad8/checks/75352180261/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error[E0164]: expected tuple struct or tuple variant, found struct variant `hir::ItemKind::Trait`
    --> compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:1110:19
     |
1110 |             kind: hir::ItemKind::Trait(.., generics, bounds, _), ..
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a tuple struct or tuple variant
     |
help: the struct variant's fields are being ignored
     |
1110 -             kind: hir::ItemKind::Trait(.., generics, bounds, _), ..
1110 +             kind: hir::ItemKind::Trait { impl_restriction: _, constness: _, is_auto: _, safety: _, ident: _, generics: _, bounds: _, items: _ }, ..
     |

For more information about this error, try `rustc --explain E0164`.
[RUSTC-TIMING] rustc_hir_analysis test:false 5.585
error: could not compile `rustc_hir_analysis` (lib) due to 1 previous error
```

</details>

fmease commented:
ready

fmease commented on compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:
```diff
@@ -902,11 +901,54 @@ impl<'a, 'tcx> Visitor<'tcx> for BoundVarContext<'a, 'tcx> {
+                self.visit_path(path, id);
+                if let Some(qself) = maybe_qself {
+                    const SELF_SEG_IDX: usize = 0;
+                    let container = self.eligible_container(path, SELF_SEG_IDX);
```
Very good catch, thanks! That's embarrassing! This regressed in the final commit. `self::Outer` vs. `Outer` also exhibits this bug, obviously.

fmease replied:
Addressed and regression test added.

lcnr commented:
I don't feel 100% confident in my rereview but remember reviewing this more indepth at the when approving the FCP. Thank you for the good work and this sure is an area of the compiler I don't want to touch myself :x

@bors r+ rollup=never

jhpratt commented:
stalled

@bors yield

@bors try jobs=dist-i686-msvc

rust-log-analyzer commented:
The job **`x86_64-gnu-miri`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27234818901/job/80424326604) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/80424326604) [(plain)](https://github.com/rust-lang/rust/commit/80a97f51412882e46a3874ad922c10bf688229ae/checks/80424326604/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass-dep/libc/libc-socketpair.rs ... ok
tests/pass-dep/libc/libc-strerror_r.rs ... ok
tests/pass-dep/libc/libc-sysconf.rs ... ok

Session terminated, killing shell...::group::Clock drift check
  local time: Tue Jun  9 21:38:39 UTC 2026
##[error]The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.
  network time: Tue, 09 Jun 2026 21:38:40 GMT
##[endgroup]
 ...killed.
```

</details>

fmease commented:
spurious @bors retry

rust-timer commented:
Finished benchmarking commit (d56483a91d6cf5041351a3208b8d08f98f0c8b56): [comparison URL](https://perf.rust-lang.org/compare.html?start=beae781308e9ddef13074a03faf57ca2fac59a5b&end=d56483a91d6cf5041351a3208b8d08f98f0c8b56&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=beae781308e9ddef13074a03faf57ca2fac59a5b&end=d56483a91d6cf5041351a3208b8d08f98f0c8b56&stat=instructions:u)
This perf run didn't have relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=beae781308e9ddef13074a03faf57ca2fac59a5b&end=d56483a91d6cf5041351a3208b8d08f98f0c8b56&stat=max-rss)
<details>
<summary>Results (secondary 3.8%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.8% | [1.4%, 6.2%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=beae781308e9ddef13074a03faf57ca2fac59a5b&end=d56483a91d6cf5041351a3208b8d08f98f0c8b56&stat=cycles:u)
This perf run didn't have relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=beae781308e9ddef13074a03faf57ca2fac59a5b&end=d56483a91d6cf5041351a3208b8d08f98f0c8b56&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0% | [0.0%, 0.1%] | 8     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.1%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.0% | [0.0%, 0.1%] | 8     |
</details>

**Bootstrap**: 523.273s -> 525.47s (0.42%)
**Artifact size**: 400.81 MiB -> 400.80 MiB (-0.00%)
